### ISSUE TITLE
Cannot connect container by its overlay network IP from host

### ISSUE THREAD

**Issue by xiangpengzhao** [moby/moby#18357]
Description of problem:
I create an overlay network and connect a container to it. I can ping the host IP from the container successfully. But it fails when I ping the container's IP which the overlay network assigned to it. 
Do I need to make more settings for the connectivity, or is that a bug or a designed behavior?

docker version

```
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64
```

docker info

```
Containers: 3
Images: 30
Server Version: 1.9.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 43
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.19.0-25-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 4
Total Memory: 3.86 GiB
Name: user11
ID: I5MJ:ANLS:EUGF:H7FV:4E3K:3KOJ:L4CD:P6LF:V3HD:AWXT:YL5E:23TJ
WARNING: No swap limit support
Cluster store: etcd://192.168.122.10:4001
Cluster advertise: 192.168.122.11:2376
```

uname -a

```
Linux user11 3.19.0-25-generic #26~14.04.1-Ubuntu SMP Fri Jul 24 21:16:20 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

Steps:
1. Create an overlay network net1 10.10.11.0/24
2. Create a container con11
3. Connect con11 to net1, the IP 10.10.11.1 is assigned to con11 
4. Ping the host's IP from con11 and ping the con11's 10.10.11.1 from host.

Expected:
ping of both direction in step 4 is successful.

Actual Results:
Ping the host's IP from con11 successfully. Ping the con11's 10.10.11.1 from host failed.

@thaJeztah @WeiZhang555

**Comment by zhaoxpZTE** [comment]
@mavenugo Would you please help me to check this issue? Thanks!

**Comment by mavenugo** [comment]
@zhaoxpZTE this is expected. Containers in overlay networks are fully isolated from being accessed from external networks (including host).

But these containers are by default given a way to reach externally via a special `docker_gwbridge` network which enables access to external network (including the host). Since these containers are also connected to the `docker_gwbridge` network, one can access the container via the interface connected to the `docker_gwbridge` network (which is eth1 inside the container).

So, instead of trying to reach the container using 10.x.x.x (assigned to eth0). Check for the IP assigned to the eth1 of the container and you can access it from the Host.

**Comment by zhaoxpZTE** [comment]
@mavenugo Thanks! I've got it.

**Comment by thaJeztah** [comment]
@mavenugo think we should improve the documentation around that?

**Comment by zhaoxpZTE** [comment]
Maybe the documentation has already mentioned that but I didn't read it carefully :sweat_smile:

**Comment by BrianAdams** [comment]
I would have expected to see a mention in http://docs.master.dockerproject.org/articles/networking/  This was helpful. Thanks!

**Comment by zhaoxpZTE** [comment]
Thanks! @BrianAdams

**Comment by prologic** [comment]
FWIW the docs I've read do mention the use of [VxLAN](https://en.wikipedia.org/wiki/Virtual_Extensible_LAN); however for those not particularly familiar it's probably worth making a little "excerpt" of the basics in the relevant docs on Overlay Networking.

**Comment by xiangpengzhao** [comment]
Thanks @prologic ! The docs are really helpful! I almost knew nothing about Overlay Networking when I opened this issue :sweat_smile:

**Comment by prologic** [comment]
Haha no worries :)


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by brugnara** [comment]
@mavenugo how can I detect the eth1 ip from outside the container? It seem that docker inspect shows only the eth0 ip. Many thanks. I've linked to the issue opened to the dokku team. Many thanks for your time.

Issue: https://github.com/dokku/dokku/issues/2085

**Comment by sanimej** [comment]
@brugnara Yes, docker inspect doesn't show the IP of the end point connected to `docker_gwbridge`. I think we should show it. Please file an issue in docker/libnetwork.

One workaround you can try is to get it from `docker network inspect docker_gwbridge`.

**Comment by josegonzalez** [comment]
Would be great if we had a canonical way of deciding how to retrieve the ip/port(s) of a container. For instance, we (dokku) currently handle the following case:

- single server, no overlay

In this case, the user wanted to do:

- single server, overlay

And I am currently investigating official docker swarm support, which requires:

- multiple servers, overlay

Ideally we have a `docker inspect` command we can run that will:

- tell us if it's an overlay network or not, and then either:
  - retrieve the non-overlay ip/port(s)
  - retrieve the overlay ip/port(s)

**Comment by josegonzalez** [comment]
Was this ever implemented? I know the original poster closed the issue, but docker/libnetwork#1082 is still open, so I assume this was not implemented either.

**Comment by xiangpengzhao** [comment]
I think I had closed this issue by mistake, sorry. Reopen for keeping track of this.

**Comment by xiangpengzhao** [comment]
I think I had closed this issue by mistake, sorry. Reopen for keeping track of this. @josegonzalez

**Comment by kor-ea** [comment]
The workaround : 
`route add -net 10.0.9.0/24 dev docker_gwbridge`

**Comment by deployable** [comment]
@kor-ea Just wanted to add a note that `ip ro add 10.0.9.0/24 dev docker_gwbridge` only works for accessing containers running on the local docker host, in case people start getting ideas.

**Comment by josegonzalez** [comment]
Should still be kept open, as the libnetwork issue wasn't fixed. Unless this is fixed elsewhere and no one was told?

**Comment by xiangpengzhao** [comment]
Okay, Jose. Since I don't follow the thread, please let me know when it's OK to close. Thanks!

**Comment by ipa-srd-rd** [comment]
Is there any possible way to communicate between one node that hosts a container in an overlay network and another container/service in the same overlay network? I can reach the local container but not the distant one. I know this is generally not possible due to security reasons but is there a security option or something else to allow it? Exposing ports wouldn't work in my case because the service is opening random ports and it is impossible to know which ones or even the range. 

I know this is not for support here but I would appreciate a hint highly! I already wrote in 4 different forums but never got a reply in any of them. 

Thanks in advance!!

**Comment by xiangpengzhao** [comment]
@ipa-srd-rd hope @mavenugo can help explain your case :)

**Comment by ipa-srd-rd** [comment]
Thanks! I've set everything up already and it works perfectly fine if I deploy the services using stack deploy on a single host but as soon as I put one of them on another host the communication stops working. I'm trying to get the Robot Operating System (ROS) to run on multiple hosts (robots) with a single centralized server, where I need to connect to each robot (service on another host/robot).

**Comment by thaJeztah** [comment]
@ipa-srd-rd check if your hosts are set up properly to allow overlay networking, as explained in the [Open protocols and ports between the hosts](https://docs.docker.com/engine/swarm/swarm-tutorial/#the-ip-address-of-the-manager-machine) section in the documentation:

> ### Open protocols and ports between the hosts
> 
> The following ports must be available. On some systems, these ports are open by default.
> 
> * **TCP port 2377** for cluster management communications
> * **TCP** and **UDP port 7946** for communication among nodes
> * **UDP port 4789** for overlay network traffic
> 
> If you are planning on creating an overlay network with encryption (`--opt encrypted`),
> you will also need to ensure **ip protocol 50** (**ESP**) traffic is allowed.

But please don't hijack tickets for unrelated issues; each comment will result in over 3000 e-mails being sent to people subscribing to this repository.

**Comment by ipa-srd-rd** [comment]
Sorry for that but I didn't find a solution after days of searching and multiple posts in different forums. Also I guess it isn't that unrelated to the overall topic in this ticket, although I am indeed aware of the fact that this is not generally the right way.

The communication within the overlay network works perfectly fine (I guess that's what the ports are for?!). Just the communication between a host and a container/service on another host doesn't work because the host itself is not part of the overlay network and, as I assume, has therefore no way to contact distant nodes in the overlay network. I am searching for a different way to contact such a container/service.

**Comment by thaJeztah** [comment]
@ipa-srd-rd this post may be helpful; https://forums.docker.com/t/how-to-check-the-network-namespace-with-overlay-networking/12395/3 (also https://www.slideshare.net/MadhuVenugopal2/dcus17-docker-networking-deep-dive, and http://techblog.d2-si.eu/2017/05/09/deep-dive-into-docker-overlay-networks-part-2.html)

**Comment by harishanand95** [comment]
I'm facing a similar issue,
I got 2 servers A and B both connected by overlay network mynet. Now I create swarm container (`--global` mode) in A and it gets assigned IP `172.18.0.5` by `docker_gwbridge` and `172.28.0.100` in mynet. I can access a service running in 172.18.0.5:80 in A but when I access the same service 172.18.0.5:80 from B, I'm not able to connect. 

Also I found that created container is listed `docker network inspect docker_gwbridge` in A and not in B.
Goal here would be to have the container services accessible at both A and B hosts. (I would prefer an option that avoids port forwarding)
